### PR TITLE
ci: ensure `Cargo.toml` and `CHANGELOG` of published crates are updated

### LINKED ISSUE #4367: ci: ensure we bump crate versions for "release worthy" changes
We have a policy to bump a crates version if we make a "release worthy" change to it. Normally, this is also accompanied with a changelog entry.

Anything that does not _exclusively_ touch the lockfile is a release worthy change to me. This for example includes https://github.com/libp2p/rust-libp2p/pull/4358.

We should have a CI check that ensures that the crate's version is bumped to `$currentPatchOnCratesIo + 1`. Such CI check would help us maintainers to remember this policy and also help users in making contributions because they can make self-directed changes based on the CI result.

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

Resolves: #4367.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -33,6 +33,8 @@ jobs:
       CRATE: ${{ matrix.crate }}
     steps:
       - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
```
Comment (thomaseizinger) [inline .github/workflows/ci.yml:37]:
```suggestion
          fetch-tags: true
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -33,6 +33,8 @@ jobs:
       CRATE: ${{ matrix.crate }}
     steps:
       - uses: actions/checkout@v4
+        with:
+          fetch-tags: true
```
Comment (thomaseizinger) [inline .github/workflows/ci.yml]:
```suggestion
          fetch-depth: 0
```

File: scripts/ensure-version-bump-and-changelog.sh [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  exit 0;
+fi
+
+# Code was touched, ensure changelog is updated too.
+if [ -z "$CHANGELOG_DIFF" ]; then
+    echo "Files in $DIR_TO_CRATE have changed, please write a changelog entry in $DIR_TO_CRATE/CHANGELOG.md"
+    exit 1
+fi
+
+# Code was touched, ensure the version used in the manifest hasn't been released yet.
```
Comment (mxinden) [inline scripts/ensure-version-bump-and-changelog.sh:33]:
```suggestion
# Code was touched, update the version used in the manifest to a version that hasn't released yet.
```

Stresses that the version needs to be changed, not that something needs to be unreleased.

File: scripts/ensure-version-bump-and-changelog.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+if git tag | grep -q "^$CRATE-v${VERSION_IN_MANIFEST}$"; then
+    echo "v$VERSION_IN_MANIFEST of '$CRATE' has already been released, please bump the version."
```
Comment (mxinden) [inline scripts/ensure-version-bump-and-changelog.sh:35]: How about linking to our docs here?

https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases

**Review by mxinden — APPROVED** [review]
Neat! As always, much appreciated Thomas!

File: scripts/ensure-version-bump-and-changelog.sh (hunk 3 above)
Comment (thomaseizinger) [reply scripts/ensure-version-bump-and-changelog.sh:33]: My comment describes what the code does, not what the user should do.

Comment (mxinden) [reply scripts/ensure-version-bump-and-changelog.sh:33]: Ah, yes. Makes sense. Sorry for the noise.

File: scripts/ensure-version-bump-and-changelog.sh (hunk 4 above)
Comment (thomaseizinger) [reply scripts/ensure-version-bump-and-changelog.sh:35]: Hmm, I think for now I'd rather keep it simple and just directly tell people here how they can fix CI.

Maybe we can just get rid of these docs in favor of the new CI checks?

Comment (mxinden) [reply scripts/ensure-version-bump-and-changelog.sh:35]: I am fine either way.

**Pull request merged**